### PR TITLE
TST: add tests to validate margin results for pivot (#25815)

### LINKED ISSUE #25815: wrong aggregated results when pivot_table index is ordered categorical data
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np

_c = np.arange(5,0,-1)

df = pd.DataFrame(np.stack([_c]*3).T, columns=['A', 'B', 'C'])

df["D"] = pd.Series(["A","B","B","A", "A"])

df["E"] = pd.cut(x=df["A"],bins=[0,3,6,10])

print(pd.pivot_table(df, index='E', columns=['D'], values="B", aggfunc="sum", margins=True))
```
``` ipython
Output is

D       A  B  All
E                
(0, 3]  3  3    9
(3, 6]  5  4    6
All     8  7   15
```
#### Problem description
the data is as follows:
``` Ipython
	A	B	C	D	E
0	5	5	5	A	(3, 6]
1	4	4	4	B	(3, 6]
2	3	3	3	B	(0, 3]
3	2	2	2	A	(0, 3]
4	1	1	1	A	(0, 3]
```
Margin calculation results are in wrong order
(0, 3] should have a sum value of 6,
(3, 6] should have a sum value of 9

If I dont use column argument, results are correct.
``` Ipython
B
E	
(0, 3]	6
(3, 6]	9
(6, 10]	0
All	15
```

I've tried both .23.4 and .24.2 . They have the same results.

#### Expected Output
``` Ipython
D       A  B  All
E                
(0, 3]  3  3    6
(3, 6]  5  4    9
All     8  7   15
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 142 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: zh_CN.GBK@cjknarrow
LOCALE: None.None

pandas: 0.23.4
pytest: 4.0.2
pip: 18.1
setuptools: 40.6.3
Cython: 0.29.2
numpy: 1.15.4
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 7.2.0
sphinx: 1.8.2
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.7
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.8
feather: None
matplotlib: 3.0.2
openpyxl: 2.5.12
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.2
lxml: 4.2.5
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: 1.2.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by peterpanmj**
- [x ] closes #25815
- [x ] tests added / passed
- [x ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/reshape/test_pivot.py [hunk 1]
```diff
@@ -286,6 +286,25 @@ def test_pivot_with_interval_index(self, interval_values, dropna):
...
+            'B': ['a', 'b', 'a', 'b'],
+            'C': pd.Categorical(ordered_cat,
+                                ordered=True).sort_values(ascending=False)
+        })
+
+        pivot_tab = pd.pivot_table(data=df, index='C', columns='B', values='A',
+                                   aggfunc='sum', margins=True)
+
+        result = pivot_tab['All']
+        expected = pivot_tab.iloc[:, :-1].sum(axis=1)
```
Comment (WillAyd) [inline pandas/tests/reshape/test_pivot.py]: Can you just construct this manually? I think generating the `result` and `expected` values from the same variable can let bugs pass through undetected

File: pandas/tests/reshape/test_pivot.py [hunk 2]
```diff
@@ -286,6 +286,25 @@ def test_pivot_with_interval_index(self, interval_values, dropna):
...
+        tm.assert_series_equal(result, expected, check_dtype=False,
```
Comment (WillAyd) [inline pandas/tests/reshape/test_pivot.py]: Can you remove `check_dtype` and `check_names`?

**Review by WillAyd — CHANGES_REQUESTED** [review]
I think this was fixed as part of #27071 - didn't see a specific test for this at first glance but if can double check would be great!

File: pandas/tests/reshape/test_pivot.py [hunk 3]
```diff
@@ -286,6 +286,25 @@ def test_pivot_with_interval_index(self, interval_values, dropna):
         expected = DataFrame({"B": 1}, index=Index(interval_values.unique(), name="A"))
         tm.assert_frame_equal(result, expected)
 
+    def test_pivot_with_interval_index_margins(self, dropna):
```
Comment (simonjayhawkins) [inline pandas/tests/reshape/test_pivot.py]: do we need to inject `dropna` here? I think you can just remove.

File: pandas/tests/reshape/test_pivot.py [hunk 4]
```diff
@@ -286,6 +286,25 @@ def test_pivot_with_interval_index(self, interval_values, dropna):
...
+        ordered_cat = pd.IntervalIndex.from_arrays(
+            [0, 0, 1, 1], [1, 1, 2, 2])
+        df = pd.DataFrame({
+            'A': np.arange(4, 0, -1),
+            'B': ['a', 'b', 'a', 'b'],
+            'C': pd.Categorical(ordered_cat,
+                                ordered=True).sort_values(ascending=False)
+        })
+
+        pivot_tab = pd.pivot_table(data=df, index='C', columns='B', values='A',
```
Comment (simonjayhawkins) [inline pandas/tests/reshape/test_pivot.py]:
```suggestion
        pivot_tab = pd.pivot_table(df, index='C', columns='B', values='A',
```

File: pandas/tests/reshape/test_pivot.py (hunk 3 above)
Comment (simonjayhawkins) [inline pandas/tests/reshape/test_pivot.py]:
```suggestion
    def test_pivot_with_interval_index_margins(self):
```

**Comment by simonjayhawkins** [comment]
see #27233 for details on back code style to resolve the CI failure.

File: pandas/tests/reshape/test_pivot.py [hunk 5]
```diff
@@ -286,6 +286,27 @@ def test_pivot_with_interval_index(self, interval_values, dropna):
         expected = DataFrame({"B": 1}, index=Index(interval_values.unique(), name="A"))
         tm.assert_frame_equal(result, expected)
 
+    def test_pivot_with_interval_index_margins(self):
+        # GH 25815
+        ordered_cat = pd.IntervalIndex.from_arrays([0, 0, 1, 1], [1, 1, 2, 2])
+        df = pd.DataFrame(
+            {
+                "A": np.arange(4, 0, -1).astype('int32'),
```
Comment (simonjayhawkins) [inline pandas/tests/reshape/test_pivot.py]:
```suggestion
                "A": np.arange(4, 0, -1, dtype=np.intp),
```

File: pandas/tests/reshape/test_pivot.py [hunk 6]
```diff
@@ -286,6 +286,27 @@ def test_pivot_with_interval_index(self, interval_values, dropna):
...
+                ),
+            }
+        )
+
+        pivot_tab = pd.pivot_table(
+            df, index="C", columns="B", values="A", aggfunc="sum", margins=True
+        )
+
+        result = pivot_tab["All"]
+        expected = pd.Series([3, 7, 10], index=result.index, name="All", dtype="int32")
```
Comment (simonjayhawkins) [inline pandas/tests/reshape/test_pivot.py]:
```suggestion
        expected = Series(
            [3, 7, 10],
            index=Index([pd.Interval(0, 1), pd.Interval(1, 2), "All"], name="C"),
            name="All",
        )
```

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
@peterpanmj changes look good. just a couple more and I think we're done. @WillAyd

File: pandas/tests/reshape/test_pivot.py (hunk 6 above)
Comment (peterpanmj) [reply pandas/tests/reshape/test_pivot.py]: Why not using index from result , seems easier ?

Comment (simonjayhawkins) [reply pandas/tests/reshape/test_pivot.py]: I'm not sure that would be in the spirit of constructing the expected result explicitly as suggested in https://github.com/pandas-dev/pandas/pull/27245#discussion_r300725193

**Review by simonjayhawkins — APPROVED** [review]

**Pull request merged**